Collected small changes and fixes

akohlmey opened the pull request:
**Summary**

This pull request combines multiple small changes and bugfixes.

**Related Issue(s)**

Addresses the issue reported here: https://matsci.org/t/python-gather-bonds-error/48618/
Fixes #3799
Addresses performance regression reported here: https://matsci.org/t/performance-degradation-in-neighbor-list-construction-in-openmp-package/48973

**Author(s)**

Axel Kohlmeyer, Temple U

**Licensing**

By submitting this pull request, I agree, that my contribution will be included in LAMMPS and redistributed under either the GNU General Public License version 2 (GPL v2) or the GNU Lesser General Public License version 2.1 (LGPL v2.1).

**Backward Compatibility**

No known issues.

**Implementation Notes**

The following individual changes are included:
- make sure the buffer for gathering bond/angle/dihedral/improper information is non-null and thus avoid a segfault
- correct minor inconsistencies in water model howto examples
- improve some error messages
- swap order of creating thermo and dump file output in order for dump styles that include thermo data to be in sync
- refactor how dump styles can access thermo data by creating an explicit cache of computed values
- add library interface API to access the cached thermo data
- fix bug when Kokkos border comm is on host
- fix performance regression when using multi-threaded neighbor list build in OPENMP package
- add support for atom type wildcards to fix reaxff/species
- use advanced fmtlib formatting to create column aligned output for fix reaxff/species instead of depending on tabs

**Post Submission Checklist**

- [ ] The feature or features in this pull request is complete
- [x] Licensing information is complete
- [x] Corresponding author information is complete
- [x] The source code follows the LAMMPS formatting guidelines
- [x] Suitable new documentation files and/or updates to the existing docs are included
- [x] The added/updated documentation is integrated and tested with the documentation build system
- [x] The feature has been verified to work with the conventional build system
- [x] The feature has been verified to work with the CMake based build system
- [x] Suitable tests have been added to the unittest tree.

akohlmey committed:
avoid null pointer dereferences by allocating a buffer for at least 1 item

akohlmey committed:
Merge branch 'develop' into collected-small-fixes

akohlmey committed:
spelling

akohlmey committed:
improve error message

akohlmey committed:
make rigid water examples more realistic and consistent. avoid warnings.

akohlmey committed:
improve error message

akohlmey committed:
improve error messages

akohlmey committed:
reorder thermo and dump output so dump styles include correct thermo data

akohlmey committed:
silence compiler warning

akohlmey committed:
fix grammar

akohlmey commented:
Note: the changes to the Output class may need extra testing. It has passed all tests I can trigger from GitHub.

athomps reviewed (approved):
I approve.

stanmoore1 committed:
Fix bug when Kokkos border comm is on host

akohlmey committed:
add cache line size padding to avoid false sharing with OPENMP package

akohlmey committed:
Merge branch 'develop' into collected-small-fixes

akohlmey committed:
add multitype data type and unittest (including tests for ubuf)

akohlmey committed:
switch dump yaml/netcdf thermo output to use new caching API. remove old API.

rbberger commented on src/NETCDF/dump_netcdf_mpiio.cpp:
```diff
@@ -594,20 +593,17 @@ void DumpNetCDFMPIIO::write()
-    for (int i = 0; i < th->nfield; i++) {
-      th->call_vfunc(i);
+    auto keywords = output->thermo->get_keywords();
+    auto fields = output->thermo->get_fields();
```
These should both be `const auto &` to avoid copies.

akohlmey replied:
Shouldn't this be automatic when Thermo::get_keywords() returns a const &?

rbberger replied:
No. That's not how `auto` works. https://godbolt.org/z/e5zPYjbdT

rbberger commented on src/NETCDF/dump_netcdf_mpiio.cpp:
```diff
@@ -422,18 +423,16 @@ void DumpNetCDFMPIIO::openfile()
-          NCERRX( ncmpi_def_var(ncid, th->keyword[i].c_str(), NC_LONG, 1, dims, &thermovar[i]), th->keyword[i].c_str() );
-#endif
+      auto fields = output->thermo->get_fields();
+      auto keywords = output->thermo->get_keywords();
```
Same here.

rbberger commented on src/NETCDF/dump_netcdf_mpiio.cpp:
```diff
@@ -318,9 +318,10 @@ void DumpNetCDFMPIIO::openfile()
-      Thermo *th = output->thermo;
-      for (int i = 0; i < th->nfield; i++) {
-        NCERRX( ncmpi_inq_varid(ncid, th->keyword[i].c_str(), &thermovar[i]), th->keyword[i].c_str() );
+      auto keywords = output->thermo->get_keywords();
```
and here

rbberger commented on src/NETCDF/dump_netcdf.cpp:
```diff
@@ -605,20 +600,17 @@ void DumpNetCDF::write()
-    for (int i = 0; i < th->nfield; i++) {
-      th->call_vfunc(i);
+    auto keywords = output->thermo->get_keywords();
+    auto fields = output->thermo->get_fields();
```
here

rbberger commented on src/NETCDF/dump_netcdf.cpp:
```diff
@@ -432,22 +433,16 @@ void DumpNetCDF::openfile()
-                               &thermovar[i]), th->keyword[i].c_str() );
-#endif
+        auto fields = output->thermo->get_fields();
+        auto keywords = output->thermo->get_keywords();
```
here

rbberger commented on src/NETCDF/dump_netcdf.cpp:
```diff
@@ -320,9 +320,10 @@ void DumpNetCDF::openfile()
-        Thermo *th = output->thermo;
-        for (int i = 0; i < th->nfield; i++) {
-          NCERRX( nc_inq_varid(ncid, th->keyword[i].c_str(), &thermovar[i]), th->keyword[i].c_str() );
+        auto keywords = output->thermo->get_keywords();
```
here

rbberger commented on src/EXTRA-DUMP/dump_yaml.cpp:
```diff
@@ -60,21 +60,26 @@ void DumpYAML::write_header(bigint ndump)
+      for (auto key : th->get_keywords()) thermo_data += fmt::format("{}, ", key);
+      thermo_data += "]\n  - data: [ ";
+
+      for (auto val : th->get_fields()) {
```
```suggestion
      for (const auto & val : th->get_fields()) {
```

rbberger commented on src/EXTRA-DUMP/dump_yaml.cpp:
```diff
@@ -60,21 +60,26 @@ void DumpYAML::write_header(bigint ndump)
+    if (update->ntimestep == th->get_timestep()) {
+
+      thermo_data += "thermo:\n  - keywords: [ ";
+      for (auto key : th->get_keywords()) thermo_data += fmt::format("{}, ", key);
```
```suggestion
      for (const auto & key : th->get_keywords()) thermo_data += fmt::format("{}, ", key);
```

rbberger reviewed (changes requested):
Nice addition, just avoid the copies.

akohlmey committed:
use explicit const references

akohlmey committed:
add accessor to nfield, make certain field types are initialized early

akohlmey committed:
make more reliable with explicit loops using exported nfield value

akohlmey committed:
Merge branch 'develop' into collected-small-fixes

akohlmey committed:
update docs

sjplimp commented:
@akohlmey New thermo field access logic looks good - thanks for iterating with me on this change

akohlmey committed:
add API to library interface to access last thermo data

akohlmey commented:
> New thermo field access logic looks good

Not quite done with that. Needed to make some tweaks so I can expose it to the library interface.
This is mostly done, but now the challenge it to get it into Python (and the other languages). To make it work with PyLammps instead of capturing and parsing the screen output, I may have to add an option to run a callback function whenever thermo output is created. But one step at a time...

rbberger commented:
@akohlmey let me know if you want me to update the Python parts.

akohlmey commented:
> @akohlmey let me know if you want me to update the Python parts.

Let me see how far I get tomorrow. Next step is writing tests (and sorting out the failing ones).
Will have to think about what a good interface for python is. Perhaps returning a dict?
Then I need to propagate the function through the other language interfaces (fortran, swig, dlopen). They are all "complete" at the moment and I would like to keep it that way, especially since we are close to a stable release.

Writing tests is a good way to check whether the added interface is usable or needs more tweaks (or adding a companion function).

akohlmey committed:
Merge branch 'develop' into collected-small-fixes

akohlmey committed:
correct multitype unittest for -DLAMMPS_SMALLSMALL

akohlmey committed:
whitespace

akohlmey committed:
add new library interface function to documentation

akohlmey committed:
allow wildcards with "cutoff" keyword to fix reaxff/species

this also switched to using fmtlib for column aligned output formatting
and re-applies clang-format.

jtclemm reviewed (approved):
Looks good.

akohlmey committed:
make sure the field_data vector size always matches the size of the keywords vector

akohlmey committed:
include versionadded tag

akohlmey committed:
search through the python folders recursive to detect all pending version tags

akohlmey committed:
add python module interface to lammps_last_thermo, small consistency fixes

akohlmey committed:
move versionadded tags to the top, replace some missed TBD with version info

akohlmey committed:
change type keyword to return a pointer to static location for better portability

akohlmey committed:
add unit tests for c-library interface and plain python module

akohlmey commented:
@rbberger test and lammps module wrapper are added. PyLammps is going to be more work, especially if you want to use this internally, so I didn't touch that. Had to make a few small tweaks to make the interface work well enough with just one function call added.

Now looking into swig, fortran, and dlopen versions...

rbberger commented:
@akohlmey looks ok. Although I personally would have stuck with the void* cast and not use the static to keep the interface "thread-safe". but I'm not even sure that's something we're striving for anyway. `typ = cast(ptr, c_int).value` should have been enough instead of resetting the function return type. Either way, I'm ok with it.

akohlmey commented:
> Although I personally would have stuck with the void* cast and not use the static to keep the interface "thread-safe".

I have tried hard and it works with ctypes, but GCC was throwing a fit when I tried to cast the void* back to an int when I was writing the unit test. The alternative would have been to use another union similar to ubuf and that would have failed on big endian machines (which are almost extinct, though) or required a maze of ifdefs to detect those. I decided that having to use a static variable in the library interface is the - by far - lesser evil.

rbberger commented:
@akohlmey Probably it's just telling us that the cleaner way would be to add a second function.

akohlmey committed:
add lammps_last_thermo support to swig, plugin and fortran interface

akohlmey committed:
document Fortran version of lammps_last_thermo
